core/parallel_state_processor.go

```diff
@@ -439,3 +444,28 @@ func GetDeps(txDependency [][]uint64) map[int][]int {
 
 	return deps
 }
+
+// returns true if dependencies are correct
+func VerifyDeps(deps map[int][]int) bool {
+	// number of transactions in the block
+	n := len(deps)
+
+	// Handle out-of-range and circular dependency problem
+	for tx, val := range deps {
+		for depTx := range val {
+			// out-of-range
+			if depTx >= n {
+				return false
+			}
```

Comment:
If we also enforce `depTx < tx` here, the circular dependency could be avoided, and we won't need the loop below. wdyt?

Reply:
Good point, Jerry. I agree.

Reply:
[Done](https://github.com/maticnetwork/bor/pull/841/commits/c7334d3f0afc99c71daa9815a31cfa8898d73d9d).